src/AutoRest.CSharp.V3/Input/Source/SourceInputModel.cs

```diff
@@ -3,31 +3,120 @@
 
 using System;
 using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
 using System.Linq;
+using Azure.Core;
+using Microsoft.CodeAnalysis;
 
 namespace AutoRest.CSharp.V3.Input.Source
 {
     public class SourceInputModel
     {
-        public SourceInputModel(
-            ModelTypeMapping[] definedSchemaTypes,
-            ClientTypeMapping[] definedClientTypes)
+        private readonly Compilation _compilation;
+        private readonly INamedTypeSymbol _schemaNameAttribute;
+        private readonly INamedTypeSymbol _clientAttribute;
+        private readonly INamedTypeSymbol _schemaMemberNameAttribute;
+
+        public SourceInputModel(Compilation compilation)
+        {
+            _compilation = compilation;
+            _schemaNameAttribute = compilation.GetTypeByMetadataName(typeof(CodeGenModelAttribute).FullName!)!;
+            _schemaMemberNameAttribute = compilation.GetTypeByMetadataName(typeof(CodeGenMemberAttribute).FullName!)!;
+            _clientAttribute = compilation.GetTypeByMetadataName(typeof(CodeGenClientAttribute).FullName!)!;
+
+            var assembly = _compilation.Assembly;
+
+            var definedSchemas = new List<ModelTypeMapping>();
+            var definedClients = new List<TypeMapping>();
+
+            foreach (IModuleSymbol module in assembly.Modules)
+            {
+                foreach (var type in GetSymbols(module.GlobalNamespace))
+                {
+                    if (type is INamedTypeSymbol namedTypeSymbol)
+                    {
+                        if (TryGetName(type, _schemaNameAttribute, out var schemaName))
+                        {
+                            var modelTypeMapping = BuildModel(schemaName, namedTypeSymbol);
+                            definedSchemas.Add(modelTypeMapping);
+                        }
+
+                        if (TryGetName(type, _clientAttribute, out var operationName))
+                        {
+                            definedClients.Add(BuildClient(operationName, namedTypeSymbol));
+                        }
+                    }
+                }
+            }
+
+            DefinedClientTypes = definedClients.ToArray();
```

Comment:
Any reason to allocate a new array and copy members? I know the properties are defined as arrays below, but why not just keep them as lists? Less time and allocation. You created the lists, so it's not like there's a shared ownership problem where copying the source list is probably a good idea.

Reply:
This is an inefficient pattern that unfortunately widely used through the codebase. I'll clean it up with a single pass in the future.

Reply:
It certainly makes sense where we need to "seal" or copy and array, though. I used it intentionally from an incoming `IEnumerable<T>` a few times in KV, but because I didn't own the enumeration and didn't want to assume anything.